### PR TITLE
DISPATCH-2238: release the action mutex before waking core thread

### PR DESCRIPTION
**Pull request by kgiusti**

### CODE REVIEW

**Review by ganeshmurthy — APPROVED** [review]

**Comment by franz1981** [comment]
@kgiusti Unrelated to this change, but it would be nice to have some flag that record the state of router thread (waiting/parked/sleeping or not) and save signaling to happen if core is already processing something.

The logic would be:

WORKER THREAD:

    signal = false
    lock 
    add action
    if (core is sleeping) 
        signal = true
    unlock
    if signal
        signal()

Core thread side we just need to mark core->sleeping right before going to sleep and it's done.
This is going to save many signaling (and talking with kernel just to inspect the state of the core thread, that's something we already know in user-space!).

**Review by ChugR — APPROVED** [review]

**Comment by kgiusti** [comment]
@franz1981 take a look at the last commit - I've implemented your suggestion.  It appears to reduce to average wait time for the action_lock, likely due to the reduction of false cond_wait() wakeups as you explained - thanks!

**Review by ChugR — CHANGES_REQUESTED** [review]

**Comment by franz1981** [comment]
> It appears to reduce to average wait time for the action_lock, likely due to the reduction of false cond_wait() wakeups as you explained - thanks!

Yep, it should help a bit if the core thread is busy processing actions, to further speed up workers to offer new ones.
In my old tests it was helping with performance too :D

File: src/router_core/router_core_thread.c [hunk 1]
```diff
@@ -213,25 +187,48 @@ void *router_core_thread(void *arg)
...
         sys_mutex_lock(core->action_lock);
 
-        //
-        // Block on the condition variable when there is no action to do
-        //
-        while (core->running && DEQ_IS_EMPTY(core->action_list)) {
-            if (!router_core_process_background_action_LH(core))
-                sys_cond_wait(core->action_cond, core->action_lock);
+        for (;;) {
+            if (!DEQ_IS_EMPTY(core->action_list)) {
```
[the remark this answers is not in the document]
Comment (kgiusti) [reply src/router_core/router_core_thread.c:191]: I originally tried exactly that - it was slightly slower than doing the check before the move.  That lock is so heavily contended that *any* extra instructions run while holding the lock results in longer acquire times by the I/O threads.

We should be good with initializing the local prior to entering the loop.

**Pull request closed without merging**